network/src/testutils/fake_socket.rs

```diff
@@ -0,0 +1,317 @@
+
+use futures::{
+    io::{AsyncRead, AsyncWrite},
+    ready,
+    task::{Context, Poll},
+};
+use memsocket::MemorySocket;
+use std::{io, pin::Pin};
+
+//
+// ReadOnlyTestSocket
+// ==================
+//
+
+pub struct ReadOnlyTestSocket<'a> {
+    /// the content
+    content: &'a [u8],
+    /// the socket will read byte-by-byte
+    fragmented: bool,
+    /// continue to read 0s once content has been fully read
+    trailing: bool,
+}
+
+impl<'a> ReadOnlyTestSocket<'a> {
+    pub fn new(content: &'a [u8]) -> Self {
+        Self {
+            content,
+            fragmented: false,
+            trailing: false,
+        }
+    }
+
+    /// reads will have to be done byte by byte
+    #[allow(dead_code)]
+    pub fn set_fragmented_read(&mut self) {
+        self.fragmented = true;
+    }
+
+    /// reads will never return pending, but 0s
+    #[allow(dead_code)]
+    pub fn set_trailing(&mut self) {
+        self.trailing = true;
+    }
+}
+
+/// Does nothing, but looks to the caller as if write worked
+impl<'a> AsyncWrite for ReadOnlyTestSocket<'a> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _context: &mut Context,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    /// Attempt to close the channel. Cannot Fail.
+    fn poll_close(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Read based on the mode set
+impl<'a> AsyncRead for ReadOnlyTestSocket<'a> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        mut _context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // nothing left to read
+        if self.content.is_empty() {
+            if self.trailing {
+                // [0, 0, 0, 0, ...]
+                for val in buf.iter_mut() {
+                    *val = 0;
+                }
+                return Poll::Ready(Ok(buf.len()));
+            } else {
+                // EOF
+                return Poll::Ready(Ok(0));
+            }
+        }
+
+        // if something left return what's asked
+        if self.fragmented {
+            // read only one byte
+            let (read, content) = self.content.split_at(1);
+            buf[0] = read[0];
+
+            // update internal state
+            self.content = content;
+
+            // return 1 byte
+            Poll::Ready(Ok(1))
+        } else {
+            // read as much as we can
+            let to_read = std::cmp::min(buf.len(), self.content.len());
+            buf[..to_read].copy_from_slice(&self.content[..to_read]);
+
+            // update internal state
+            self.content = &self.content[to_read..self.content.len()];
+
+            // return length read
+            Poll::Ready(Ok(to_read))
+        }
+    }
+}
+
+//
+// ExposingTestSocket
```

Comment:
Just wondering, what does "ExposingTestSocket" mean?  it's a test socket that shows what's on the socket?

Reply:
Better name indeed! I'll rename